src/fsharp/FSharp.Core/array.fs

```diff
@@ -494,12 +494,23 @@ namespace Microsoft.FSharp.Collections
 
         [<CompiledName("Filter")>]
         let filter f (array: _[]) = 
-            checkNonNull "array" array
-            let res = List<_>() // ResizeArray
```

Comment:
Only issue I can think of with this is if you were already close to OOM and you do a filter on a large array, this implementation is more likely to OOM since it will always allocate a large array; the original allocated more conservatively (at the cost of multiple copies and re-allocations). Not a common scenario though.